GH-3672: Revert serialization but put restriction in place.

mpapierski opened the pull request:
Closes #3672

This reverts bytesrepr and bincode (and serde in general) serialization formats back to before #3219 was merged. But to satisfy the new requirement of configurable vesting schedules, a restriction is put in place so the length would not exceed fixed-sized slots of 14 weeks.

Commit:
Revert serialization but put restriction in place.

This reverts bytesrepr and bincode (and serde in general) serialization
formats back, but to satisfy new requirement of configurable vesting
schedules, a restriction is put in place so the length would not exceed
fixed sized slots.

Fraser999 commented on node/src/types/chainspec/core_config.rs:
```diff
@@ -14,10 +14,25 @@ use serde::{
 };
 use tracing::{error, warn};
 
+fn vesting_schedule_period<'de, D: Deserializer<'de>>(
```
Let's stick with validating in the `is_valid` method please.  Something like:
```rust
        if self.vesting_schedule_period.millis() > VESTING_SCHEDULE_LENGTH_MILLIS {
            error!(
                vesting_schedule_millis = self.vesting_schedule_period.millis(),
                max_millis = VESTING_SCHEDULE_LENGTH_MILLIS
                "vesting schedule period too long"
            );
            return false;
        }
```

mpapierski replied:
Addressed

Fraser999 commented on types/src/system/auction/bid/vesting.rs:
```diff
@@ -74,23 +67,31 @@ impl VestingSchedule {
         let locked_amounts_length =
             vesting_schedule_period_to_weeks(vesting_schedule_period_millis);
 
+        if locked_amounts_length as usize >= LOCKED_AMOUNTS_MAX_LENGTH {
+            // We will not initialize a vesting schedule if the genesis configuration is incorrect.
+            // There are restrictions put in place in the chainspec config to make sure we will not
+            // hit this case.
+            //
+            // This is done to avoid breaking changes in the serialization format.
+            return false;
+        }
```
I think we should panic here, since returning `false` doesn't really convey the error.
```suggestion
        assert!(
            locked_amounts_length < LOCKED_AMOUNTS_LENGTH as u64,
            "vesting schedule period must be less than {} weeks",
            LOCKED_AMOUNTS_LENGTH
        );
```

Let's also add the following to the doc comment:
```rust
    ///
    /// # Panics
    ///
    /// Panics if `vesting_schedule_period_millis` represents more than 13 weeks.
```

mpapierski replied:
Addressed

Fraser999 commented on types/src/system/auction/bid/vesting.rs:
```diff
@@ -151,104 +153,87 @@ impl VestingSchedule {
-            }
+impl FromBytes for [U512; LOCKED_AMOUNTS_MAX_LENGTH] {
+    fn from_bytes(mut bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let mut result: MaybeUninit<[U512; LOCKED_AMOUNTS_MAX_LENGTH]> = MaybeUninit::uninit();
```
Let's avoid `MaybeUninit`.

mpapierski replied:
Done, and moved the serialization logic to `fn write_bytes`

Fraser999 commented on execution_engine_testing/test_support/src/chainspec_config.rs:
```diff
@@ -39,9 +39,20 @@ pub enum Error {
-    FailedToCreateGenesisRequest,
+}
+
+fn vesting_schedule_period<'de, D: Deserializer<'de>>(
```
Again, let's not do this to validate config settings.  This should be changed to match how we do it in the "real" config, ideally sharing code if possible.

mpapierski replied:
Ticketed the refactor to share most code (# to be announced). For now, refactored just enough to always call the new method `fn is_valid()` before returning a valid `ChainspecConfig` instance.

Commit:
Address @Fraser999's comments

Commit:
Fix tests

One of the methods still assumed the locked amounts is a vec, so this
commit is finding a zero in the sequence, and trims the amounts only to
these that are nonzero.

rafal-ch reviewed (approved):
LGTM :+1:

mpapierski commented:
bors r+
